### PR TITLE
doc: declarations and attributes

### PR DESCRIPTION
**Pull request by EdAyers**
More docstrings relating to decls

### CODE REVIEW

File: src/Lean/Declaration.lean [hunk 1]
```diff
@@ -192,13 +192,23 @@ def Declaration.isUnsafeInductiveDeclEx : Declaration → Bool
...
+  numIndices : Nat
+  /-- List of all (including this one) inductive datatypes in the mutual declaration containing this one -/
+  all : List Name
+  /-- List of the names of the constructors for this inductive datatype. -/
+  ctors : List Name
+  /-- `true` when recursive (that is, the inductive type appears as an argument in a constructor). -/
+  isRec : Bool
+  /-- Whether the definition is flagged as unsafe. -/
   isUnsafe : Bool
   isReflexive : Bool
```
Comment (EdAyers) [inline src/Lean/Declaration.lean:209]: What does isReflexive do?

File: src/Lean/Declaration.lean [hunk 2]
```diff
@@ -226,10 +236,14 @@ def mkInductiveValEx (name : Name) (levelParams : List Name) (type : Expr) (numP
 def InductiveVal.numCtors (v : InductiveVal) : Nat := v.ctors.length
 
 structure ConstructorVal extends ConstantVal where
-  induct  : Name    -- Inductive Type this Constructor is a member of
-  cidx    : Nat     -- Constructor index (i.e., Position in the inductive declaration)
-  numParams : Nat   -- Number of parameters in inductive datatype `induct`
-  numFields : Nat   -- Number of fields (i.e., arity - nparams)
+  /-- Inductive Type this Constructor is a member of -/
```
Comment (Kha) [inline src/Lean/Declaration.lean:239]:
```suggestion
  /-- Inductive type this constructor is a member of -/
```
Any time you see weird casing, it's probably from our semi-automated naming style change :)

File: src/Lean/Elab/DeclModifiers.lean [hunk 3]
```diff
@@ -26,6 +26,7 @@ def checkNotAlreadyDeclared {m} [Monad m] [MonadEnv m] [MonadError m] (declName
     if env.contains declName then
       throwError "a non-private declaration '{declName}' has already been declared"
 
+/-- Whether a declaration is regular, protected or private. -/
```
Comment (Kha) [inline src/Lean/Elab/DeclModifiers.lean:29]: Does this kind of doc comment really add value to the declaration? How about "Declaration visibility modifier" in this case?

File: src/Lean/ResolveName.lean [hunk 4]
```diff
@@ -216,6 +219,26 @@ def resolveGlobalConstNoOverloadCore [Monad m] [MonadResolveName m] [MonadEnv m]
   | [c] => pure c
   | _   => throwError s!"ambiguous identifier '{mkConst n}', possible interpretations: {cs.map mkConst}"
 
+/-- Interpret the syntax `n` as an identifier for a global constant, and return a list of resolved
+constant names that it could be refering to based on the currently open namespaces.
+This should be used instead of `resolveGlobalConstCore` because `Syntax` objects may have names that have already been resolved.
```
Comment (Kha) [inline src/Lean/ResolveName.lean:224]: This drops the "for identifiers taken from syntax" specifier from above

File: src/Lean/Elab/DeclModifiers.lean (hunk 3 above)
Comment (EdAyers) [reply src/Lean/Elab/DeclModifiers.lean:29]: I think it does because you get this information when you hover it without needing to go-to-def.

**Review by Kha — APPROVED** [review]

**Pull request merged**